File: api/src/test/java/org/apache/gravitino/semantic/TestSemanticModelMembers.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.semantic;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.apache.gravitino.NameIdentifier;
import org.junit.jupiter.api.Test;

public class TestSemanticModelMembers {

  @Test
  public void testBuildMembers() {
    Expression expression = expression("order_amount");
    AIContext aiContext = AIContext.of("Use certified fields");
    CustomExtension extension =
        CustomExtension.builder().withVendorName("example").withData("{\"key\":\"value\"}").build();
    Field field =
        Field.builder()
            .withName("order_amount")
            .withExpression(expression)
            .withDimension(Dimension.builder().withIsTime(false).build())
            .withLabel("Order amount")
            .withDescription("Gross order amount")
            .withDatatype(DataType.DECIMAL)
            .withAIContext(aiContext)
            .withCustomExtensions(new CustomExtension[] {extension})
            .build();
    Dataset dataset =
        Dataset.builder()
            .withName("orders")
            .withSource(NameIdentifier.of("sales", "mart", "orders"))
            .withPrimaryKey(new String[] {"order_id"})
            .withUniqueKeys(new String[][] {{"order_id"}})
            .withDescription("Orders")
            .withAIContext(aiContext)
            .withFields(new Field[] {field})
            .withCustomExtensions(new CustomExtension[] {extension})
            .build();
    Relationship relationship =
        Relationship.builder()
            .withName("orders_to_customers")
            .withFrom("orders")
            .withTo("customers")
            .withFromColumns(new String[] {"customer_id"})
            .withToColumns(new String[] {"id"})
            .withAIContext(aiContext)
            .build();
    Metric metric =
        Metric.builder()
            .withName("total_revenue")
            .withExpression(expression)
            .withDescription("Total revenue")
            .withDatatype(DataType.DECIMAL)
            .withAIContext(aiContext)
            .build();

    assertEquals("orders", dataset.name());
    assertEquals(NameIdentifier.of("sales", "mart", "orders"), dataset.source());
    assertEquals("order_amount", dataset.fields()[0].name());
    assertEquals("orders", relationship.from());
    assertEquals("customers", relationship.to());
    assertEquals(DataType.DECIMAL, metric.datatype());
  }

  @Test
  public void testCollectionsAreDefensivelyCopied() {
    Expression expression = expression("order_id");
    String[] primaryKey = {"order_id"};
    String[][] uniqueKeys = {{"order_id"}};
    Field[] fields = {Field.builder().withName("order_id").withExpression(expression).build()};
    Dataset dataset =
        Dataset.builder()
            .withName("orders")
            .withSource(NameIdentifier.of("sales", "mart", "orders"))
            .withPrimaryKey(primaryKey)
            .withUniqueKeys(uniqueKeys)
            .withFields(fields)
            .build();
    String[] fromColumns = {"customer_id"};
    String[] toColumns = {"id"};
    Relationship relationship =
        Relationship.builder()
            .withName("orders_to_customers")
            .withFrom("orders")
            .withTo("customers")
            .withFromColumns(fromColumns)
            .withToColumns(toColumns)
            .build();

    primaryKey[0] = "changed";
    uniqueKeys[0][0] = "changed";
    fields[0] = Field.builder().withName("changed").withExpression(expression).build();
    fromColumns[0] = "changed";
    toColumns[0] = "changed";

    assertArrayEquals(new String[] {"order_id"}, dataset.primaryKey());
    assertArrayEquals(new String[] {"order_id"}, dataset.uniqueKeys()[0]);
    assertEquals("order_id", dataset.fields()[0].name());
    assertArrayEquals(new String[] {"customer_id"}, relationship.fromColumns());
    assertArrayEquals(new String[] {"id"}, relationship.toColumns());
    assertNotSame(dataset.uniqueKeys(), dataset.uniqueKeys());
    assertNotSame(dataset.uniqueKeys()[0], dataset.uniqueKeys()[0]);
  }

  @Test
  public void testOptionalCollectionsPreserveNullAndEmpty() {
    Expression expression = expression("id");
    Field unsetField = Field.builder().withName("id").withExpression(expression).build();
    Field emptyField =
        Field.builder()
            .withName("id")
            .withExpression(expression)
            .withCustomExtensions(new CustomExtension[0])
            .build();
    Dataset unsetDataset = datasetBuilder().build();
    Dataset emptyDataset =
        datasetBuilder()
            .withPrimaryKey(new String[0])
            .withUniqueKeys(new String[0][])
            .withFields(new Field[0])
            .withCustomExtensions(new CustomExtension[0])
            .build();
    Relationship unsetRelationship = relationshipBuilder().build();
    Relationship emptyRelationship =
        relationshipBuilder().withCustomExtensions(new CustomExtension[0]).build();
    Metric unsetMetric = Metric.builder().withName("count").withExpression(expression).build();
    Metric emptyMetric =
        Metric.builder()
            .withName("count")
            .withExpression(expression)
            .withCustomExtensions(new CustomExtension[0])
            .build();

    assertNull(unsetDataset.primaryKey());
    assertNull(unsetDataset.uniqueKeys());
    assertNull(unsetDataset.fields());
    assertNull(unsetDataset.customExtensions());
    assertNull(unsetField.customExtensions());
    assertNull(unsetRelationship.customExtensions());
    assertNull(unsetMetric.customExtensions());

    assertArrayEquals(new String[0], emptyDataset.primaryKey());
    assertEquals(0, emptyDataset.uniqueKeys().length);
    assertArrayEquals(new Field[0], emptyDataset.fields());
    assertArrayEquals(new CustomExtension[0], emptyDataset.customExtensions());
    assertArrayEquals(new CustomExtension[0], emptyField.customExtensions());
    assertArrayEquals(new CustomExtension[0], emptyRelationship.customExtensions());
    assertArrayEquals(new CustomExtension[0], emptyMetric.customExtensions());

    assertNotEquals(unsetDataset, emptyDataset);
    assertNotEquals(unsetField, emptyField);
    assertNotEquals(unsetRelationship, emptyRelationship);
    assertNotEquals(unsetMetric, emptyMetric);
  }

  @Test
  public void testRequiredBuilderFields() {
    assertThrows(IllegalArgumentException.class, () -> Field.builder().build());
    assertThrows(IllegalArgumentException.class, () -> Dataset.builder().build());
    assertThrows(IllegalArgumentException.class, () -> Relationship.builder().build());
    assertThrows(IllegalArgumentException.class, () -> Metric.builder().build());
  }

  @Test
  public void testArrayElementValidation() {
    Expression expression = expression("id");

    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withPrimaryKey(new String[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withPrimaryKey(new String[] {""}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withUniqueKeys(new String[][] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withUniqueKeys(new String[][] {new String[0]}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withUniqueKeys(new String[][] {{null}}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withFields(new Field[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> datasetBuilder().withCustomExtensions(new CustomExtension[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            Field.builder()
                .withName("id")
                .withExpression(expression)
                .withCustomExtensions(new CustomExtension[] {null})
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            Metric.builder()
                .withName("count")
                .withExpression(expression)
                .withCustomExtensions(new CustomExtension[] {null})
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () -> relationshipBuilder().withFromColumns(new String[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> relationshipBuilder().withToColumns(new String[] {""}).build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            relationshipBuilder()
                .withFromColumns(new String[] {"tenant_id", "customer_id"})
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () -> relationshipBuilder().withCustomExtensions(new CustomExtension[] {null}).build());
  }

  private static Dataset.Builder datasetBuilder() {
    return Dataset.builder()
        .withName("orders")
        .withSource(NameIdentifier.of("sales", "mart", "orders"));
  }

  private static Relationship.Builder relationshipBuilder() {
    return Relationship.builder()
        .withName("orders_to_customers")
        .withFrom("orders")
        .withTo("customers")
        .withFromColumns(new String[] {"customer_id"})
        .withToColumns(new String[] {"id"});
  }

  private static Expression expression(String value) {
    return Expression.builder()
        .withDialects(
            new DialectExpression[] {
              DialectExpression.builder()
                  .withDialect(Dialects.ANSI_SQL)
                  .withExpression(value)
                  .build()
            })
        .build();
  }
}

File: clients/client-python/tests/unittests/api/semantic/test_semantic_model_members.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import unittest

from gravitino.api.semantic.ai_context import AIContext
from gravitino.api.semantic.custom_extension import CustomExtension
from gravitino.api.semantic.data_type import DataType
from gravitino.api.semantic.dataset import Dataset
from gravitino.api.semantic.dialect_expression import DialectExpression
from gravitino.api.semantic.dialects import Dialects
from gravitino.api.semantic.dimension import Dimension
from gravitino.api.semantic.expression import Expression
from gravitino.api.semantic.field import Field
from gravitino.api.semantic.metric import Metric
from gravitino.api.semantic.relationship import Relationship
from gravitino.name_identifier import NameIdentifier


def _expression(expression: str = "order_amount") -> Expression:
    return Expression([DialectExpression(Dialects.ANSI_SQL, expression)])


class TestSemanticModelMembers(unittest.TestCase):
    def test_field_defaults(self):
        field = Field("order_amount", _expression())

        self.assertEqual("order_amount", field.name())
        self.assertEqual(_expression(), field.expression())
        self.assertIsNone(field.dimension())
        self.assertIsNone(field.label())
        self.assertIsNone(field.description())
        self.assertIsNone(field.datatype())
        self.assertIsNone(field.ai_context())
        self.assertIsNone(field.custom_extensions())

    def test_field_with_all_members(self):
        extension = CustomExtension("acme", "{}")
        field = Field(
            name="order_date",
            expression=_expression("order_date"),
            dimension=Dimension(is_time=True),
            label="Order date",
            description="The date the order was placed",
            datatype=DataType.DATE,
            ai_context=AIContext.of("Order placement date"),
            custom_extensions=[extension],
        )

        self.assertEqual(Dimension(is_time=True), field.dimension())
        self.assertEqual("Order date", field.label())
        self.assertEqual("The date the order was placed", field.description())
        self.assertEqual(DataType.DATE, field.datatype())
        self.assertEqual(AIContext.of("Order placement date"), field.ai_context())
        self.assertEqual([extension], field.custom_extensions())

    def test_field_equality(self):
        field = Field("order_amount", _expression(), datatype=DataType.DECIMAL)
        equal_field = Field("order_amount", _expression(), datatype=DataType.DECIMAL)

        self.assertEqual(equal_field, field)
        self.assertEqual(hash(equal_field), hash(field))
        self.assertNotEqual(Field("order_amount", _expression()), field)
        self.assertNotEqual(field, "invalid")

    def test_field_rejects_invalid_arguments(self):
        with self.assertRaisesRegex(ValueError, "name must not be null or empty"):
            Field("", _expression())
        with self.assertRaisesRegex(ValueError, "expression must not be null"):
            Field("order_amount", None)
        with self.assertRaisesRegex(
            ValueError, r"customExtensions\[0\] must not be null"
        ):
            Field("order_amount", _expression(), custom_extensions=[None])

    def test_dataset_defaults(self):
        source = NameIdentifier.of("sales", "mart", "orders")
        dataset = Dataset("orders", source)

        self.assertEqual("orders", dataset.name())
        self.assertEqual(source, dataset.source())
        self.assertIsNone(dataset.primary_key())
        self.assertIsNone(dataset.unique_keys())
        self.assertIsNone(dataset.description())
        self.assertIsNone(dataset.ai_context())
        self.assertIsNone(dataset.fields())
        self.assertIsNone(dataset.custom_extensions())

    def test_dataset_with_all_members(self):
        field = Field("order_amount", _expression())
        dataset = Dataset(
            name="orders",
            source=NameIdentifier.of("sales", "mart", "orders"),
            primary_key=["order_id"],
            unique_keys=[["order_id"], ["customer_id", "order_date"]],
            description="Order facts",
            ai_context=AIContext.of("Certified order facts"),
            fields=[field],
            custom_extensions=[CustomExtension("acme", "{}")],
        )

        self.assertEqual(["order_id"], dataset.primary_key())
        self.assertEqual(
            [["order_id"], ["customer_id", "order_date"]], dataset.unique_keys()
        )
        self.assertEqual("Order facts", dataset.description())
        self.assertEqual([field], dataset.fields())

    def test_dataset_returns_copies(self):
        primary_key = ["order_id"]
        unique_keys = [["order_id"]]
        fields = [Field("order_amount", _expression())]
        dataset = Dataset(
            "orders",
            NameIdentifier.of("sales", "mart", "orders"),
            primary_key=primary_key,
            unique_keys=unique_keys,
            fields=fields,
        )

        primary_key.append("mutated")
        unique_keys[0].append("mutated")
        fields.clear()
        dataset.primary_key().clear()
        dataset.unique_keys()[0].clear()
        dataset.fields().clear()

        self.assertEqual(["order_id"], dataset.primary_key())
        self.assertEqual([["order_id"]], dataset.unique_keys())
        self.assertEqual(1, len(dataset.fields()))

    def test_dataset_rejects_invalid_arguments(self):
        source = NameIdentifier.of("sales", "mart", "orders")

        with self.assertRaisesRegex(ValueError, "name must not be null or empty"):
            Dataset("", source)
        with self.assertRaisesRegex(ValueError, "source must not be null"):
            Dataset("orders", None)
        with self.assertRaisesRegex(
            ValueError, r"primaryKey\[0\] must not be null or empty"
        ):
            Dataset("orders", source, primary_key=[""])
        with self.assertRaisesRegex(
            ValueError, r"uniqueKeys\[0\] must not be null or empty"
        ):
            Dataset("orders", source, unique_keys=[[]])
        with self.assertRaisesRegex(
            ValueError, r"uniqueKeys\[1\]\[0\] must not be null or empty"
        ):
            Dataset("orders", source, unique_keys=[["order_id"], [None]])
        with self.assertRaisesRegex(ValueError, r"fields\[0\] must not be null"):
            Dataset("orders", source, fields=[None])

    def test_metric(self):
        metric = Metric(
            name="total_revenue",
            expression=_expression("SUM(orders.order_amount)"),
            description="Total revenue across all orders",
            datatype=DataType.DECIMAL,
        )

        self.assertEqual("total_revenue", metric.name())
        self.assertEqual(_expression("SUM(orders.order_amount)"), metric.expression())
        self.assertEqual("Total revenue across all orders", metric.description())
        self.assertEqual(DataType.DECIMAL, metric.datatype())
        self.assertIsNone(metric.ai_context())
        self.assertIsNone(metric.custom_extensions())

    def test_metric_equality(self):
        metric = Metric("total_revenue", _expression())
        equal_metric = Metric("total_revenue", _expression())

        self.assertEqual(equal_metric, metric)
        self.assertEqual(hash(equal_metric), hash(metric))
        self.assertNotEqual(Metric("other", _expression()), metric)
        self.assertNotEqual(metric, "invalid")

    def test_metric_rejects_invalid_arguments(self):
        with self.assertRaisesRegex(ValueError, "name must not be null or empty"):
            Metric("", _expression())
        with self.assertRaisesRegex(ValueError, "expression must not be null"):
            Metric("total_revenue", None)

    def test_relationship(self):
        relationship = Relationship(
            name="orders_to_customers",
            from_dataset="orders",
            to_dataset="customers",
            from_columns=["customer_id"],
            to_columns=["id"],
        )

        self.assertEqual("orders_to_customers", relationship.name())
        self.assertEqual("orders", relationship.from_dataset())
        self.assertEqual("customers", relationship.to_dataset())
        self.assertEqual(["customer_id"], relationship.from_columns())
        self.assertEqual(["id"], relationship.to_columns())
        self.assertIsNone(relationship.ai_context())
        self.assertIsNone(relationship.custom_extensions())

    def test_relationship_returns_copies(self):
        from_columns = ["customer_id"]
        relationship = Relationship(
            "orders_to_customers", "orders", "customers", from_columns, ["id"]
        )

        from_columns.append("mutated")
        relationship.from_columns().clear()

        self.assertEqual(["customer_id"], relationship.from_columns())

    def test_relationship_equality(self):
        relationship = Relationship(
            "orders_to_customers", "orders", "customers", ["customer_id"], ["id"]
        )
        equal_relationship = Relationship(
            "orders_to_customers", "orders", "customers", ["customer_id"], ["id"]
        )

        self.assertEqual(equal_relationship, relationship)
        self.assertEqual(hash(equal_relationship), hash(relationship))
        self.assertNotEqual(relationship, "invalid")

    def test_relationship_rejects_invalid_arguments(self):
        with self.assertRaisesRegex(ValueError, "name must not be null or empty"):
            Relationship("", "orders", "customers", ["customer_id"], ["id"])
        with self.assertRaisesRegex(ValueError, "from must not be null or empty"):
            Relationship("rel", "", "customers", ["customer_id"], ["id"])
        with self.assertRaisesRegex(ValueError, "to must not be null or empty"):
            Relationship("rel", "orders", "", ["customer_id"], ["id"])
        with self.assertRaisesRegex(
            ValueError, "fromColumns must not be null or empty"
        ):
            Relationship("rel", "orders", "customers", [], ["id"])
        with self.assertRaisesRegex(ValueError, "toColumns must not be null or empty"):
            Relationship("rel", "orders", "customers", ["customer_id"], [])
        with self.assertRaisesRegex(
            ValueError, r"fromColumns\[0\] must not be null or empty"
        ):
            Relationship("rel", "orders", "customers", [""], ["id"])
        with self.assertRaisesRegex(
            ValueError, "fromColumns and toColumns must have the same length"
        ):
            Relationship("rel", "orders", "customers", ["customer_id"], ["id", "extra"])
